dask/task_spec.py

```diff
@@ -0,0 +1,742 @@
+
+""" Task specification for dask
+
+This module contains the task specification for dask. It is used to represent
+runnable and non-runnable (i.e. literals) tasks in a dask graph.
+
+Simple examples of how to express tasks in dask
+-----------------------------------------------
+
+.. code-block:: python
+
+    func("a", "b") ~ Task("key", func, ["a", "b"])
+
+    [func("a"), func("b")] ~ SequenceOfTasks(
+        "key",
+        [Task("key-1", func, "a"), Task("key-2", func, "b")]
+    )
+
+    {"a": func("b")} ~ DictOfTasks("key", {"a": Task("a", func, "b")})
+
+    "literal-string" ~ Literal("key", "literal-string")
+
+
+Keys, Aliases and KeyRefs
+-------------------------
+
+Keys are used to identify tasks in a dask graph. Every `BaseTask` instance has a
+key attribute that _should_ reference the key in the dask graph.
+
+.. code-block:: python
+
+    {"key": Task("key", func, "a")}
+
+Referencing other tasks is possible by using either one of `Alias` or a
+`KeyRef`. (both can be used interchangeably; The KeyRef is a proxy for the
+distributed WrappedKey subclass and is used here as a stand-in replacement. The
+implementation deals with both transparently).
+
+.. code-block:: python
+
+    # KeyRef can be used to provide the name of the reference explicitly
+    t = Task("key", func, KeyRef("key-1"))
+
+    # If a task is still in scope, the method `ref` can be used for convenience
+    t2 = Task("key2", func2, t.ref())
+
+Nested tasks and inlining
+-------------------------
+
+Tasks can be nested in a dict or list. This is useful for expressing more complex tasks.
+
+.. code-block:: python
+
+    func(func2("a", "b"), "c") ~ Task("key", func, [Task("key-1", func2, "a", "b"), "c"])
+
+    {"a": func("b")} ~ DictOfTasks("key", {"a": Task("a", func, "b")})
+
+This kind of nesting is often a byproduct of an optimizer step that inlines the
+tasks explicitly. This is done by calling the `inline` method on a task.
+
+.. code-block:: python
+
+    t1 = Task("key-1", func, "a")
+    t2 = Task("key-2", func, KeyRef("key-1"))
+
+    t2.inline({"key-1": t1}) ~ Task("key-2", func, Task("key-1", func, "a"))
+
+Executing a task
+----------------
+
+A task can be executed by calling it with a dictionary of values. The values
+should contain the dependencies of the task.
+
+.. code-block:: python
+
+    t = Task("key", add, [KeyRef("a"), KeyRef("b")])
+    assert t.dependencies == {"a", "b"}
+    t({"a": 1, "b": 2}) == 3
+
+"""
+
+import itertools
+import pickle
+from collections import defaultdict
+from collections.abc import Mapping
+from typing import TYPE_CHECKING, Any, Callable, overload
+
+import cloudpickle
+
+from dask.base import normalize_token
+from dask.core import reverse_dict
+from dask.optimization import SubgraphCallable
+from dask.typing import Key as KeyType
+from dask.utils import funcname, is_namedtuple_instance
+
+if TYPE_CHECKING:
+    from typing import TypeVar
+
+    _T = TypeVar("_T", bound="BaseTask")
+
+
+class WrappedKey:
```

Comment:
This had to move from distributed since we'd get otherwise horrible circular imports